core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java

```diff
@@ -104,41 +116,85 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
-                        indexMetaData.getIndex());
-                    newIndices.put(indexMetaData.getIndex(), indexMetaData);
+            Map<Index, IndexMetaData> delIndices = new HashMap<>();
+            for (IndexStateMetaData persistedIndex : indexMetaDataList) {
+                if (persistedIndex != null) {
+                    final Index index = persistedIndex.getIndexMetaData().getIndex();
+                    final String clusterUUID = metaData.clusterUUID();
+                    if (persistedIndex.isSameClusterUUID(clusterUUID)) {
+                        // the cluster metadata's cluster UUID is the same as the cluster UUID that created this index,
+                        // so the index shouldn't be imported as dangling but rather deleted and cleaned up, because the
+                        // dangling files are probably due to remnants from deleted indices (e.g. the node was offline while
+                        // the index was deleted) or failed cleanup
+                        logger.info("[{}] dangling index exists on the local file system, but it is deleted in the metadata of the " +
+                                    "same cluster that created it, deleting the index on the local file system", index);
+                        delIndices.put(index, persistedIndex.getIndexMetaData());
+                    } else if (metaData.hasIndex(index.getName())) {
+                        logger.warn("[{}] can not be imported as a dangling index, as index with same name already exists " +
+                                    "in cluster metadata", index);
+                    } else {
+                        logger.info("[{}] dangling index, exists on local file system, but not in cluster metadata, " +
+                                    "auto import to cluster state", index);
+                        newIndices.put(index, persistedIndex.getIndexMetaData());
+                    }
                 }
             }
-            return newIndices;
+            return Tuple.tuple(newIndices, delIndices);
         } catch (IOException e) {
             logger.warn("failed to list dangling indices", e);
-            return emptyMap();
+            return Tuple.tuple(emptyMap(), emptyMap());
+        }
+    }
+
+    /**
+     * Cleans the indices that exist on the local file system but should be deleted according to the metadata state
+     * and we have decided that the indices should not be imported as dangling.
+     */
+    void cleanDeletedIndices(final Map<Index, IndexMetaData> deletedIndices, final ClusterState clusterState) {
+        for (Map.Entry<Index, IndexMetaData> entry : deletedIndices.entrySet()) {
+            final String indexName = entry.getKey().getName();
+            final IndexMetaData indexMetaData = entry.getValue();
+            try {
+                indicesService.deleteIndexStore("index deleted but metadata still on file system", indexMetaData, clusterState, false);
```

Comment:
I think we should use:
```java
nodeEnv.deleteIndexDirectorySafe(indexMetaData.getIndex(), 0, new IndexSettings(indexMetaData, settings));
```
`indicesService.deleteIndexStore` additionally checks the cluster metadata to make sure we can delete the index store, IMO this check doesn't make sense here as we have already read the index in question as dangled indices w.r.t. the cluster metadata?